File: web/js/pages/broker.js
// 메시지 브로커(RabbitMQ · Kafka) 관리 화면.
//
// DB 화면과 나눈 이유: 여기에는 테이블도 SQL도 없다. 대신 큐·토픽·컨슈머 그룹이 있고,
// 운영자가 이 화면에서 답을 얻고 싶은 질문은 셋이다 — 브로커는 괜찮은가, 메시지가
// 쌓이고 있나, 그것을 줄이는 소비자가 있나.
//
// 스토리지 화면과 나눈 이유: 스토리지의 첫 질문은 "얼마나 찼나"지만, 브로커의 첫 질문은
// "쌓이고 있나"다. 큐 깊이·랙은 그 자체로 좋지도 나쁘지도 않다 — 소비자가 따라오고
// 있으면 정상이고, 따라오지 못하면 장애의 전조다.
import { api } from '../core/api.js';
import { state, kindLabel } from '../core/store.js';
import {
  h, mount, icon, select, spinner, emptyState, pageHeader,
  badge, envBadge, formatDate, relativeTime, toast, toastError, confirmDialog, openModal,
} from '../core/ui.js';
import { formatBytes } from '../core/chart.js';
import { navigate } from '../core/router.js';
import { errorPanel } from './users.js';

export async function renderBroker(outlet, params, query) {
  mount(outlet, spinner('커넥션 목록을 불러오는 중…'));

  let conns;
  try {
    conns = await api.get('/connections/');
  } catch (err) {
    mount(outlet, errorPanel(err));
    return;
  }

  // 브로커 능력이 켜진 종류만 고른다. 종류 이름을 화면에서 나열하면
  // 종류가 늘 때마다 여기도 고쳐야 한다.
  const usable = conns.items.filter((i) => i.accessible
    && state.meta?.dbKinds?.find((k) => k.kind === i.connection.kind)?.capabilities?.broker);

  if (usable.length === 0) {
    mount(outlet,
      pageHeader('메시지 브로커', 'RabbitMQ · Kafka 클러스터 관리'),
      emptyState('등록된 메시지 브로커가 없습니다. RabbitMQ 또는 Kafka를 커넥션으로 등록하세요.',
        h('a.btn', { href: '/connections' }, 'DB 커넥션으로')),
    );
    return;
  }

  const selectedId = query.get('conn') || usable[0].connection.id;
  const current = usable.find((i) => i.connection.id === selectedId) ?? usable[0];
  const conn = current.connection;

  const connSelect = select(
    usable.map((i) => ({
      value: i.connection.id,
      label: `${i.connection.name} — ${kindLabel(i.connection.kind)}`,
    })),
    { value: conn.id },
  );
  connSelect.addEventListener('change', () => {
    navigate(`/broker?conn=${encodeURIComponent(connSelect.value)}`);
  });

  const body = h('div');

  mount(outlet,
    pageHeader('메시지 브로커', `${conn.name} 클러스터를 관리합니다`, [
      h('button.btn', { type: 'button', onclick: () => load() }, icon('refresh'), '다시 읽기'),
    ]),
    h('div.card.filter-bar', {},
      h('label.field.field-inline', {}, h('span.field-label', {}, '브로커'), connSelect),
      envBadge(conn.environment),
      h('div.filter-sep'),
      h('a.btn.btn-small', { href: `/monitor?conn=${encodeURIComponent(conn.id)}` },
        icon('activity'), '모니터링'),
      h('a.btn.btn-small', { href: `/events?conn=${encodeURIComponent(conn.id)}` },
        icon('alert'), '이벤트'),
    ),
    body,
  );

  async function load() {
    mount(body, spinner(`${conn.name} 상태를 읽는 중…`));
    try {
      const res = await api.get(`/connections/${conn.id}/broker`);
      mount(body, ...clusterView(conn, res));
    } catch (err) {
      mount(body, errorPanel(err));
    }
  }
  await load();
}

// clusterView는 개요 + 종류별 목록이다.
function clusterView(conn, res) {
  const ov = res.overview ?? {};
  const feats = res.features ?? {};
  const out = [overviewCard(ov)];

  if (feats.queues) out.push(queuesCard(conn, feats.write));
  if (feats.exchanges) out.push(exchangesCard(conn));
  if (feats.connections) out.push(connectionsCard(conn, feats.write));
  if (feats.topics) out.push(topicsCard(conn));
  if (feats.groups) out.push(groupsCard(conn));
  return out;
}

function healthBadge(health) {
  const level = health?.level ?? 'unknown';
  const label = health?.summary || level;
  if (level === 'ok') return badge(label, 'success');
  if (level === 'warn') return badge(label, 'warn');
  if (level === 'error') return badge(label, 'danger');
  return badge(label, 'neutral');
}

function overviewCard(ov) {
  return h('div.card', {},
    h('div.card-title', {},
      h('span', {}, '브로커 상태'),
      healthBadge(ov.health),
      ov.version ? h('span.muted.small', {}, ov.version) : null,
    ),
    // 적체가 먼저다. 브로커에서 가장 자주 확인하는 것이고, 소비자가 따라오지
    // 못하는 순간부터는 다른 모든 지표보다 급한 문제가 된다.
    h('div.storage-capacity', {},
      h('div.storage-capacity-text', {},
        h('strong', {}, `밀린 메시지 ${ov.backlog?.toLocaleString('ko-KR') ?? 0}개`),
        h('span.muted', {}, `소비자 ${ov.consumers?.toLocaleString('ko-KR') ?? 0}명이 처리 중`),
      ),
    ),
    (ov.health?.checks ?? []).length
      ? h('ul.storage-checks', {}, ov.health.checks.map((c) => h('li', {}, c)))
      : null,
    h('div.storage-facts', {}, (ov.facts ?? []).map((f) =>
      h('div.storage-fact', { class: f.level ? `is-${f.level}` : '' },
        h('span.storage-fact-label', {}, f.label),
        h('span.storage-fact-value', {}, f.value)))),
    (ov.notes ?? []).length
      ? h('div.storage-notes', {}, ov.notes.map((n) => h('p.field-help', {}, n)))
      : null,
  );
}

// ---------- RabbitMQ 큐 ----------

function queuesCard(conn, canWrite) {
  const card = h('div.card');
  const draw = async () => {
    mount(card, h('h2.card-title', {}, '큐'), spinner('목록을 읽는 중…'));
    try {
      const res = await api.get(`/connections/${conn.id}/broker/queues`);
      const queues = res.queues ?? [];
      mount(card,
        h('div.card-title', {}, h('span', {}, '큐'),
          h('span.muted.small', {}, `${queues.length}개`)),
        queues.length === 0
          ? emptyState('큐가 없습니다.')
          : h('table.table', {},
            h('thead', {}, h('tr', {},
              h('th', {}, '이름'), h('th', {}, '상태'), h('th', {}, '대기'),
              h('th', {}, '미확인'), h('th', {}, '소비자'), h('th', {}, '초당 발행 · 소비'),
              h('th', {}, '메모리'), canWrite ? h('th', {}, '') : null)),
            h('tbody', {}, queues.map((q) => h('tr', { class: q.starved ? 'is-starved' : '' },
              h('td', {}, h('span', { title: `vhost ${q.vhost}` }, q.name)),
              h('td', {}, queueState(q.state)),
              h('td', {}, q.ready.toLocaleString('ko-KR')),
              h('td', {}, q.unacked.toLocaleString('ko-KR')),
              h('td', {}, q.consumers.toLocaleString('ko-KR')),
              h('td', {}, `${q.publishRate?.toFixed(1) ?? '-'} · ${q.deliverRate?.toFixed(1) ?? '-'}`),
              h('td', {}, q.memory ? formatBytes(q.memory) : '-'),
              canWrite
                ? h('td.row-actions', {},
                  h('button.icon-btn', {
                    type: 'button', title: '메시지 비우기',
                    onclick: () => confirmPurge(conn, q, draw),
                  }, icon('trash', 13)),
                  h('button.icon-btn', {
                    type: 'button', title: '큐 삭제',
                    onclick: () => confirmDeleteQueue(conn, q, draw),
                  }, icon('x', 13)),
                )
                : null,
            )))),
      );
    } catch (err) {
      mount(card, h('h2.card-title', {}, '큐'), errorPanel(err));
    }
  };
  draw();
  return card;
}

function queueState(s) {
  if (s === 'running') return badge('정상', 'success');
  if (s === 'flow' || s === 'blocked') return badge(s, 'warn');
  return badge(s ?? '-', 'neutral');
}

async function confirmPurge(conn, q, done) {
  const ok = await confirmDialog({
    title: '메시지 비우기',
    message: `${q.name} 큐의 메시지 ${q.total.toLocaleString('ko-KR')}개를 모두 버립니다. 되돌릴 수 없습니다.`,
    confirmLabel: '비우기',
    danger: true,
    requireText: q.name,
  });
  if (!ok) return;
  try {
    await api.post(`/connections/${conn.id}/broker/purge`, { vhost: q.vhost, name: q.name });
    toast('큐를 비웠습니다', 'success');
    done();
  } catch (err) {
    toastError(err);
  }
}

async function confirmDeleteQueue(conn, q, done) {
  const ok = await confirmDialog({
    title: '큐 삭제',
    message: `${q.name} 큐를 지웁니다. 바인딩과 메시지가 함께 사라지며 되돌릴 수 없습니다.`,
    confirmLabel: '삭제',
    danger: true,
    requireText: q.name,
  });
  if (!ok) return;
  try {
    await api.post(`/connections/${conn.id}/broker/delete-queue`, { vhost: q.vhost, name: q.name });
    toast('큐를 삭제했습니다', 'success');
    done();
  } catch (err) {
    toastError(err);
  }
}

// ---------- RabbitMQ 익스체인지 ----------

function exchangesCard(conn) {
  const card = h('div.card');
  const draw = async () => {
    mount(card, h('h2.card-title', {}, '익스체인지'), spinner('목록을 읽는 중…'));
    try {
      const res = await api.get(`/connections/${conn.id}/broker/exchanges`);
      const exchanges = res.exchanges ?? [];
      mount(card,
        h('div.card-title', {}, h('span', {}, '익스체인지'),
          h('span.muted.small', {}, `${exchanges.length}개`)),
        exchanges.length === 0
          ? emptyState('익스체인지가 없습니다.')
          : h('table.table', {},
            h('thead', {}, h('tr', {},
              h('th', {}, '이름'), h('th', {}, '종류'), h('th', {}, '지속성'),
              h('th', {}, '초당 들어옴 · 나감'))),
            h('tbody', {}, exchanges.map((e) => h('tr', {},
              h('td', {}, e.name),
              h('td', {}, h('code', {}, e.type)),
              h('td', {}, e.durable ? badge('durable', 'neutral') : badge('일시', 'warn')),
              h('td', {}, `${e.inRate?.toFixed(1) ?? '-'} · ${e.outRate?.toFixed(1) ?? '-'}`),
            )))),
      );
    } catch (err) {
      mount(card, h('h2.card-title', {}, '익스체인지'), errorPanel(err));
    }
  };
  draw();
  return card;
}

// ---------- RabbitMQ 연결 ----------

function connectionsCard(conn, canWrite) {
  const card = h('div.card');
  const draw = async () => {
    mount(card, h('h2.card-title', {}, '클라이언트 연결'), spinner('목록을 읽는 중…'));
    try {
      const res = await api.get(`/connections/${conn.id}/broker/connections`);
      const connections = res.connections ?? [];
      mount(card,
        h('div.card-title', {}, h('span', {}, '클라이언트 연결'),
          h('span.muted.small', {}, `${connections.length}개`)),
        connections.length === 0
          ? emptyState('연결된 클라이언트가 없습니다.')
          : h('table.table', {},
            h('thead', {}, h('tr', {},
              h('th', {}, '사용자'), h('th', {}, 'vhost'), h('th', {}, '상태'),
              h('th', {}, '채널'), h('th', {}, '피어'), h('th', {}, '프로토콜'),
              h('th', {}, '연결 시각'), canWrite ? h('th', {}, '') : null)),
            h('tbody', {}, connections.map((c) => h('tr', {},
              h('td', {}, c.user),
              h('td', {}, c.vhost),
              h('td', {}, c.state),
              h('td', {}, c.channels),
              h('td', {}, c.peer),
              h('td', {}, c.protocol || '-'),
              h('td', {}, c.connectedAt && !c.connectedAt.startsWith('0001')
                ? h('span', { title: formatDate(c.connectedAt) }, relativeTime(c.connectedAt)) : '-'),
              canWrite
                ? h('td.row-actions', {},
                  h('button.icon-btn', {
                    type: 'button', title: '연결 끊기',
                    onclick: () => confirmCloseConnection(conn, c, draw),
                  }, icon('x', 13)),
                )
                : null,
            )))),
      );
    } catch (err) {
      mount(card, h('h2.card-title', {}, '클라이언트 연결'), errorPanel(err));
    }
  };
  draw();
  return card;
}

async function confirmCloseConnection(conn, c, done) {
  const ok = await confirmDialog({
    title: '연결 끊기',
    message: `${c.user} (${c.peer}) 연결을 끊습니다. 클라이언트는 즉시 재연결을 시도할 수 있습니다.`,
    confirmLabel: '끊기',
    danger: true,
  });
  if (!ok) return;
  try {
    await api.post(`/connections/${conn.id}/broker/close-connection`, { name: c.name });
    toast('연결을 끊었습니다', 'success');
    done();
  } catch (err) {
    toastError(err);
  }
}

// ---------- Kafka 토픽 ----------

function topicsCard(conn) {
  const card = h('div.card');
  const draw = async () => {
    mount(card, h('h2.card-title', {}, '토픽'), spinner('목록을 읽는 중…'));
    try {
      const res = await api.get(`/connections/${conn.id}/broker/topics?limit=200`);
      const topics = res.topics ?? [];
      mount(card,
        h('div.card-title', {}, h('span', {}, '토픽'),
          h('span.muted.small', {}, `${topics.length}개`)),
        topics.length === 0
          ? emptyState('토픽이 없습니다.')
          : h('table.table', {},
            h('thead', {}, h('tr', {},
              h('th', {}, '이름'), h('th', {}, '파티션'), h('th', {}, '복제'),
              h('th', {}, '보관 메시지'), h('th', {}, '복제 부족'), h('th', {}, '오프라인'))),
            h('tbody', {}, topics.map((t) => h('tr', {},
              h('td', {},
                t.internal
                  ? h('span', { title: '카프카 내부 토픽' }, t.name)
                  : h('button.link-btn', {
                    type: 'button',
                    onclick: () => openTopicConfig(conn, t),
                  }, t.name)),
              h('td', {}, t.partitions),
              h('td', {}, t.replicationFactor),
              h('td', {}, t.messages < 0 ? '-' : t.messages.toLocaleString('ko-KR')),
              h('td', {}, t.underReplicated
                ? badge(t.underReplicated, 'warn') : h('span.muted', {}, '0')),
              h('td', {}, t.offline
                ? badge(t.offline, 'danger') : h('span.muted', {}, '0')),
            )))),
      );
    } catch (err) {
      mount(card, h('h2.card-title', {}, '토픽'), errorPanel(err));
    }
  };
  draw();
  return card;
}

// openTopicConfig는 토픽 설정을 모달로 보여준다.
async function openTopicConfig(conn, t) {
  const body = h('div');
  openModal({
    title: t.name,
    width: 560,
    body: () => [body],
  });
  mount(body, spinner('설정을 읽는 중…'));
  try {
    const res = await api.get(`/connections/${conn.id}/broker/topics/${encodeURIComponent(t.name)}/config`);
    const entries = Object.entries(res.config ?? {});
    mount(body,
      h('p.field-help', {}, `파티션 ${t.partitions}개 · 복제 ${t.replicationFactor} · `
        + `보관 메시지 ${t.messages < 0 ? '알 수 없음' : t.messages.toLocaleString('ko-KR')}개`),
      entries.length === 0
        ? emptyState('설정이 없습니다.')
        : h('table.table', {},
          h('thead', {}, h('tr', {}, h('th', {}, '키'), h('th', {}, '값'))),
          h('tbody', {}, entries.map(([k, v]) => h('tr', {},
            h('td', {}, h('code', {}, k)),
            h('td', {}, v || h('span.muted', {}, '(기본값)')))))),
    );
  } catch (err) {
    mount(body, errorPanel(err));
  }
}

// ---------- Kafka 컨슈머 그룹 ----------

function groupsCard(conn) {
  const card = h('div.card');
  const draw = async () => {
    mount(card, h('h2.card-title', {}, '컨슈머 그룹'), spinner('목록을 읽는 중…'));
    try {
      const res = await api.get(`/connections/${conn.id}/broker/groups`);
      const groups = res.groups ?? [];
      mount(card,
        h('div.card-title', {}, h('span', {}, '컨슈머 그룹'),
          h('span.muted.small', {}, `${groups.length}개`)),
        groups.length === 0
          ? emptyState('컨슈머 그룹이 없습니다.')
          : h('table.table', {},
            h('thead', {}, h('tr', {},
              h('th', {}, '이름'), h('th', {}, '상태'), h('th', {}, '멤버'),
              h('th', {}, '랙'), h('th', {}, '토픽별 랙'))),
            h('tbody', {}, groups.map((g) => h('tr', { class: g.lag > 0 && g.members === 0 ? 'is-starved' : '' },
              h('td', {}, g.name),
              h('td', {}, groupState(g.state)),
              h('td', {}, g.members),
              h('td', {}, g.lag > 0
                ? badge(g.lag.toLocaleString('ko-KR'), g.members === 0 ? 'warn' : 'neutral')
                : h('span.muted', {}, '0')),
              h('td', {}, (g.topics ?? []).slice(0, 3).map((tl) =>
                h('span.muted.small', { style: 'margin-right:8px' },
                  `${tl.topic} ${tl.lag.toLocaleString('ko-KR')}`)),
              ),
            )))),
      );
    } catch (err) {
      mount(card, h('h2.card-title', {}, '컨슈머 그룹'), errorPanel(err));
    }
  };
  draw();
  return card;
}

function groupState(s) {
  if (s === 'Stable') return badge('정상', 'success');
  if (s === 'Empty') return badge('비어 있음', 'neutral');
  if (s === 'Dead') return badge('죽음', 'danger');
  if (s === 'PreparingRebalance' || s === 'CompletingRebalance' || s === 'Assigning') {
    return badge('재조정 중', 'warn');
  }
  return badge(s ?? '-', 'neutral');
}

File: internal/broker/broker.go
// Package broker는 메시지 브로커(RabbitMQ·Kafka)를 다룬다.
//
// 스토리지와 같은 자리에 두지 않은 이유: 물어보는 것이 다르다. 스토리지에서는 "얼마나 찼나"가
// 첫 질문이지만, 브로커에서는 **"쌓이고 있나"** 가 첫 질문이다. 큐 깊이는 그 자체로는 좋지도
// 나쁘지도 않다 — 소비자가 따라오고 있으면 정상이고, 따라오지 못하면 장애의 전조다.
// 그래서 이 패키지는 언제나 "쌓인 양"과 "그것을 줄이는 쪽"을 함께 본다.
//
// 두 브로커의 접근 방법이 다르다.
//   - RabbitMQ: 관리 플러그인의 HTTP API(기본 15672)를 쓴다.
//   - Kafka: 브로커에 HTTP API가 없다. 카프카 프로토콜을 직접 말하는 순수 Go 클라이언트
//     (franz-go)로 메타데이터·컨슈머 랙을 읽는다. REST 프록시를 요구하면 그것을 따로 세워야
//     하는 배포에서는 이 기능을 아예 쓸 수 없다.
package broker

import (
	"context"
	"time"

	"dbstudio/internal/opsapi"
)

// 이 패키지가 다루는 종류.
const (
	KindRabbitMQ = "rabbitmq"
	KindKafka    = "kafka"
)

// 공통 어휘는 opsapi에 있다(storage와 같은 이유로 별칭을 둔다).
type (
	Health = opsapi.Health
	Fact   = opsapi.Fact
	Config = opsapi.Config
)

const (
	HealthOK      = opsapi.HealthOK
	HealthWarn    = opsapi.HealthWarn
	HealthError   = opsapi.HealthError
	HealthUnknown = opsapi.HealthUnknown
)

// ConfigFrom은 커넥션과 자격증명에서 접속 설정을 만든다.
var ConfigFrom = opsapi.ConfigFrom

// Overview는 개요 화면 한 장이다.
type Overview struct {
	Kind    string `json:"kind"`
	Version string `json:"version,omitempty"`
	Health  Health `json:"health"`
	// Backlog는 지금 브로커에 남아 있는 메시지 수다(RabbitMQ는 큐의 합, Kafka는 총 랙).
	//
	// 두 브로커의 서로 다른 개념을 같은 칸에 담는 이유: 운영자가 화면을 열고 처음 보는
	// 숫자는 "밀린 게 얼마나 되나"이고, 그 답은 브로커 종류와 무관하게 같은 뜻이다.
	Backlog int64 `json:"backlog"`
	// Consumers는 그 적체를 줄이고 있는 쪽의 수다(소비자 또는 컨슈머 그룹 멤버).
	Consumers int64    `json:"consumers"`
	Facts     []Fact   `json:"facts"`
	Notes     []string `json:"notes,omitempty"`
}

// Queue는 RabbitMQ 큐 하나다.
type Queue struct {
	Name  string `json:"name"`
	VHost string `json:"vhost"`
	State string `json:"state"`
	// Ready는 소비자에게 아직 나가지 않은 메시지, Unacked는 나갔지만 확인되지 않은 메시지다.
	Ready       int64   `json:"ready"`
	Unacked     int64   `json:"unacked"`
	Total       int64   `json:"total"`
	Consumers   int64   `json:"consumers"`
	Memory      int64   `json:"memory"`
	Node        string  `json:"node,omitempty"`
	Durable     bool    `json:"durable"`
	PublishRate float64 `json:"publishRate"`
	DeliverRate float64 `json:"deliverRate"`
	IdleSince   string  `json:"idleSince,omitempty"`
	// Starved는 "메시지는 쌓여 있는데 소비자가 없다"는 뜻이다. 브로커에서 가장 흔한
	// 사고 유형이고, 큐 목록에서 눈에 먼저 들어와야 한다.
	Starved bool `json:"starved"`
}

// Exchange는 RabbitMQ 익스체인지 하나다.
type Exchange struct {
	Name    string `json:"name"`
	VHost   string `json:"vhost"`
	Type    string `json:"type"`
	Durable bool   `json:"durable"`
	// InRate/OutRate는 들어오고 나가는 초당 메시지다. 라우팅이 끊긴 익스체인지는
	// 들어오는 것만 있고 나가는 것이 없다.
	InRate  float64 `json:"inRate"`
	OutRate float64 `json:"outRate"`
}

// NodeInfo는 브로커 노드 하나다(RabbitMQ 노드 또는 Kafka 브로커).
type NodeInfo struct {
	Name    string `json:"name"`
	Running bool   `json:"running"`
	// Host/Port는 Kafka 브로커에 쓴다.
	Host string `json:"host,omitempty"`
	Port int    `json:"port,omitempty"`
	// Controller/Leader 표시(Kafka).
	Controller bool `json:"controller,omitempty"`

	MemUsed   int64    `json:"memUsed,omitempty"`
	MemLimit  int64    `json:"memLimit,omitempty"`
	DiskFree  int64    `json:"diskFree,omitempty"`
	DiskLimit int64    `json:"diskLimit,omitempty"`
	FDUsed    int64    `json:"fdUsed,omitempty"`
	FDTotal   int64    `json:"fdTotal,omitempty"`
	Alarms    []string `json:"alarms,omitempty"`
	// Partitions는 네트워크 분단으로 갈라져 보이는 다른 노드들이다(RabbitMQ).
	// 비어 있지 않으면 클러스터가 쪼개진 상태다.
	Partitions []string `json:"partitions,omitempty"`
	Uptime     int64    `json:"uptime,omitempty"`
}

// Topic은 Kafka 토픽 하나다.
type Topic struct {
	Name              string `json:"name"`
	Partitions        int    `json:"partitions"`
	ReplicationFactor int    `json:"replicationFactor"`
	// UnderReplicated는 ISR이 복제 수를 못 채운 파티션 수다.
	UnderReplicated int `json:"underReplicated"`
	// Offline은 리더가 없는 파티션 수다. 그 파티션은 읽지도 쓰지도 못한다.
	Offline int `json:"offline"`
	// Messages는 (끝 오프셋 - 시작 오프셋)의 합이다. 정확한 메시지 수가 아니라
	// **보관 중인 양의 근사**다 — 압축(compaction)과 보관 기간 삭제 때문이다.
	Messages int64 `json:"messages"`
	Internal bool  `json:"internal"`
	// Config는 눈에 띄어야 하는 설정만 담는다(보관 기간, 정리 정책 등).
	Config map[string]string `json:"config,omitempty"`
}

// Group은 컨슈머 그룹 하나다(Kafka).
type Group struct {
	Name     string `json:"name"`
	State    string `json:"state"`
	Members  int    `json:"members"`
	Protocol string `json:"protocol,omitempty"`
	// Lag은 이 그룹이 아직 읽지 못한 메시지 수의 합이다. 브로커 운영에서 가장 중요한 숫자다.
	Lag    int64      `json:"lag"`
	Topics []GroupLag `json:"topics,omitempty"`
}

// GroupLag는 그룹이 구독 중인 토픽 하나의 랙이다.
type GroupLag struct {
	Topic string `json:"topic"`
	Lag   int64  `json:"lag"`
}

// Conn은 브로커에 붙어 있는 클라이언트 연결 하나다(RabbitMQ).
type Conn struct {
	Name        string    `json:"name"`
	User        string    `json:"user"`
	VHost       string    `json:"vhost"`
	State       string    `json:"state"`
	Channels    int       `json:"channels"`
	Peer        string    `json:"peer"`
	Protocol    string    `json:"protocol,omitempty"`
	ConnectedAt time.Time `json:"connectedAt,omitempty"`
}

// Client는 두 브로커가 공통으로 답하는 것이다.
type Client interface {
	Kind() string
	Overview(ctx context.Context) (*Overview, error)
}

// Metrics는 폴러가 쓰는 값이다.
//
// 두 브로커의 지표를 한 구조로 모으는 이유: 임계값 룰은 종류를 모르고도 걸려야 한다.
// "밀린 메시지"와 "그것을 줄이는 쪽의 수"는 어느 브로커에나 있다.
type Metrics struct {
	Health Health
	// Backlog는 밀린 메시지(RabbitMQ: ready+unacked, Kafka: 총 랙).
	Backlog int64
	// Unacked는 나갔지만 확인되지 않은 메시지다(RabbitMQ).
	Unacked int64
	// Consumers는 소비자 수(RabbitMQ) 또는 그룹 멤버 수(Kafka).
	Consumers int64
	// Starved는 "메시지는 있는데 소비자가 없는" 큐·그룹의 수다.
	Starved int64
	// Queues/Topics/Groups/Nodes는 규모다.
	Queues    int64
	Topics    int64
	Groups    int64
	Nodes     int64
	NodesDown int64
	// PublishRate/DeliverRate는 초당 메시지다(RabbitMQ).
	PublishRate float64
	DeliverRate float64
	// MaxLag는 그룹 하나가 가진 가장 큰 랙이다(Kafka).
	MaxLag int64
	// Offline/UnderReplicated는 파티션 상태다(Kafka).
	Offline         int64
	UnderReplicated int64
	// Alarms는 메모리·디스크 알람 수다(RabbitMQ).
	Alarms int64
}
